bigquery_etl/backfill/shredder_mitigation.py

```diff
@@ -0,0 +1,196 @@
+    date_formats = [
+        ("%Y-%m-%d", date),
+        ("%Y-%m-%d %H:%M:%S", datetime),
+        ("%Y-%m-%dT%H:%M:%S", datetime),
+        ("%Y-%m-%dT%H:%M:%SZ", datetime),
+        ("%Y-%m-%d %H:%M:%S UTC", datetime),
+        ("%H:%M:%S", time),
+    ]
+    for format, dtype in date_formats:
+        try:
+            if dtype == time:
+                parsed_to_time = datetime.strptime(value, format).time()
+                if isinstance(parsed_to_time, time):
+                    return DataTypeGroup.DATE
+            else:
+                parsed_to_date = datetime.strptime(value, format)
+                if isinstance(parsed_to_date, dtype):
+                    return DataTypeGroup.DATE
+        except (ValueError, TypeError):
+            continue
+    if isinstance(value, time):
+        return DataTypeGroup.DATE
+    if isinstance(value, date):
+        return DataTypeGroup.DATE
+    elif isinstance(value, bool):
+        return DataTypeGroup.BOOLEAN
+    if isinstance(value, int):
+        return DataTypeGroup.NUMERIC
+    elif isinstance(value, float):
+        return DataTypeGroup.FLOAT
+    elif isinstance(value, str) or isinstance(value, bytes):
+        return DataTypeGroup.STRING
+    elif isinstance(value, NoneType):
+        return DataTypeGroup.UNDETERMINED
+    else:
+        raise ValueError(f"Unsupported data type: {type(value)}")
+
+
+def classify_columns(
+    new_row: dict, existing_columns: list, new_columns: list
+) -> tuple[list[Column], list[Column], list[Column], list[Column], list[Column]]:
+    """Compare the new row with the existing columns and return the list of common, added and removed columns by type."""
+    common_dimensions = []
+    added_dimensions = []
+    removed_dimensions = []
+    metrics = []
+    undefined = []
+
+    for key in existing_columns:
+        if key in existing_columns and key not in new_columns:
+            removed_dimensions.append(
+                Column(
+                    key,
+                    DataTypeGroup.UNDETERMINED,
+                    ColumnType.UNDETERMINED,
+                    ColumnStatus.REMOVED,
+                )
+            )
+
+    for key, value in new_row.items():
+        if key in existing_columns:
+            common_dimensions.append(
+                Column(
+                    key,
+                    get_bigquery_type(value),
+                    ColumnType.DIMENSION,
+                    ColumnStatus.COMMON,
+                )
+            )
+        elif key not in existing_columns and key in new_columns:
+            added_dimensions.append(
+                Column(
+                    key,
+                    get_bigquery_type(value),
+                    ColumnType.DIMENSION,
+                    ColumnStatus.ADDED,
+                )
+            )
+        elif (
+            key not in existing_columns
+            and key not in new_columns
+            and get_bigquery_type(value) is DataTypeGroup.NUMERIC
```

Comment:
you could use `:=` to save the result for `get_bigquery_type(value)` so you don't have to call the function again